### PR TITLE
ENH: suports wiring up q2-diversity's alpha, beta, a/b_phylo, and a/b_rarefaction

### LINKED ISSUE #22: ENH: Provide dispatch functions for q2-diversity
**Improvement Description**
`diversity-lib` should provide functions that dispatch calls from `diversity alpha` and `diversity beta` to the best extant implementation of the selected measure.

### PR DESCRIPTION
**Pull request by ChrisKeefe**
Closes #22, required by https://github.com/qiime2/q2-diversity/issues/280
Supersedes and closes #21 
See also https://github.com/qiime2/q2-diversity/pull/281

54702b4  closes https://github.com/qiime2/q2-diversity/issues/286
54702b4  closes https://github.com/qiime2/q2-diversity/issues/288
4aa2ff77 closes https://github.com/qiime2/q2-diversity/issues/255

### CODE REVIEW

File: q2_diversity_lib/plugin_setup.py [hunk 1]
```diff
@@ -243,3 +250,100 @@
...
+    input_descriptions={'table': "The feature table containing the samples "
+                        "for which a selected metric should be computed."},
+    parameter_descriptions={'metric':
+                            'The alpha diversity metric to be computed.'},
+    output_descriptions={'vector': "Vector containing per-sample values "
+                                   "for the chosen metric."},
+    name="Alpha Passthrough (non-phylogenetic)",
+    description="Computes a vector of values for each samples in a feature "
+                "table using the scikit-bio implementation of a chosen alpha "
+                "diversity metric."
```
Comment (thermokarst) [inline q2_diversity_lib/plugin_setup.py:269]: This makes it sound like the return data is many vectors (one for each sample), which isn't the case.

```suggestion
    description="Computes a vector of values (one value for each sample in a feature "
                "table) using the scikit-bio implementation of the selected alpha "
                "diversity metric."
```

File: q2_diversity_lib/plugin_setup.py [hunk 2]
```diff
@@ -243,3 +250,100 @@
...
+        'metric': 'The beta diversity metric to be computed.',
+        'pseudocount': ('A pseudocount to handle zeros for compositional '
+                        'metrics.  This is ignored for other metrics.'),
+        'n_jobs': n_jobs_description
+    },
+    output_descriptions={'distance_matrix': 'The resulting distance matrix.'},
+    name='Beta Passthrough (non-phylogenetic)',
+    description=("Computes a distance matrix for all pairs of samples in a "
+                 "feature table using the scikit-bio implementation of a "
+                 "chosen beta diversity metric."),
```
Comment (thermokarst) [inline q2_diversity_lib/plugin_setup.py:294]:
```suggestion
    description=("Computes a distance matrix for all pairs of samples in a "
                 "feature table using the scikit-bio implementation of the "
                 "selected beta diversity metric."),
```

File: q2_diversity_lib/plugin_setup.py [hunk 3]
```diff
@@ -243,3 +250,100 @@
...
+plugin.methods.register_function(
+    function=q2_diversity_lib.beta_passthrough,
+    inputs={'table': FeatureTable[Frequency]},
+    parameters={'metric': Str % Choices(beta.METRICS['NONPHYLO']['UNIMPL']),
+                'pseudocount': Int % Range(1, None),
+                'n_jobs': Int % Range(1, None) | Str % Choices(['auto'])},
+    outputs=[('distance_matrix', DistanceMatrix)],
+    input_descriptions={
+        'table': ('The feature table containing the samples over which beta '
+                  'diversity should be computed.')
```
Comment (thermokarst) [inline q2_diversity_lib/plugin_setup.py:282]: No need for the superfluous parenthesis.

```suggestion
        'table': 'The feature table containing the samples over which beta '
                 'diversity should be computed.'
```

I think in general you have gone with the paren-free style in most of the other actions defined in here, so please scour and remove all unnecessary parens.

File: q2_diversity_lib/plugin_setup.py [hunk 4]
```diff
@@ -243,3 +250,100 @@
...
+    },
+    parameter_descriptions={
+        'metric': 'The beta diversity metric to be computed.',
+        'pseudocount': ('A pseudocount to handle zeros for compositional '
+                        'metrics.  This is ignored for other metrics.'),
```
Comment (thermokarst) [inline q2_diversity_lib/plugin_setup.py:287]:
```suggestion
        'pseudocount': 'A pseudocount to handle zeros for compositional '
                       'metrics. This is ignored for non-compositional metrics.',
```

File: q2_diversity_lib/tests/test_alpha.py [hunk 5]
```diff
@@ -15,23 +15,28 @@
...
 from q2_types.tree import NewickFormat
 from q2_diversity_lib import (faith_pd, pielou_evenness, observed_features,
-                              shannon_entropy)
+                              shannon_entropy, alpha, translate_metric_name)
+from qiime2 import Artifact
 
-nonphylogenetic_measures = [observed_features, pielou_evenness,
-                            shannon_entropy]
+nonphylogenetic_metrics = alpha.METRICS['NONPHYLO']['IMPL']
+metric_name_translations = alpha.METRICS['METRIC_NAME_TRANSLATIONS']
```
Comment (thermokarst) [inline q2_diversity_lib/tests/test_alpha.py:22]: These are only used one time - I don't think it is worth polluting the namespace with these new variables. Please remove these.

`nonphylogenetic_metrics` is only 4 chars less than `METRICS['NONPHYLO']['IMPL']`, and if you drop the extra `METRIC_` in `METRIC_NAME_TRANSLATIONS` you'll also only be saving yourself 4 chars, not worth it IMO to mostly just add an extra layer of indirection.

File: q2_diversity_lib/tests/test_alpha.py [hunk 6]
```diff
@@ -15,23 +15,28 @@
...
+                                                empty_table)
 
     def test_non_phylogenetic_passed_empty_table(self):
-        for measure in nonphylogenetic_measures:
+        for metric in nonphylogenetic_metrics:
+            metric_tr = translate_metric_name(metric, metric_name_translations)
+            method = self.plugin.actions[metric_tr]
             with self.assertRaisesRegex(ValueError, 'empty'):
-                measure(table=self.empty_table)
+                method(table=self.empty_table)
```
Comment (thermokarst) [inline q2_diversity_lib/tests/test_alpha.py:39]:
```suggestion
                self.plugin.actions[metric_tr](table=self.empty_table)
```

This still comes in under 80 cols, and would allow you to remove line 37 above.

File: q2_diversity_lib/tests/test_alpha.py [hunk 7]
```diff
@@ -15,23 +15,28 @@
...
         super().setUp()
-        self.empty_table = biom.Table(np.array([]), [], [])
+        empty_table = biom.Table(np.array([]), [], [])
+        self.empty_table = Artifact.import_data('FeatureTable[Frequency]',
+                                                empty_table)
 
     def test_non_phylogenetic_passed_empty_table(self):
-        for measure in nonphylogenetic_measures:
+        for metric in nonphylogenetic_metrics:
+            metric_tr = translate_metric_name(metric, metric_name_translations)
```
Comment (thermokarst) [inline q2_diversity_lib/tests/test_alpha.py:36]: This doesn't seem worth the extra indirection - why not just keep all of the "logic" in one place - the `METRICS` data structure defined in the module?

File: q2_diversity_lib/tests/test_alpha.py [hunk 8]
```diff
@@ -234,3 +239,42 @@ def test_do_not_drop_undefined_samples(self):
...
         pdt.assert_series_equal(actual, self.expected)
+
+
+class AlphaPassthroughTests(TestPluginBase):
+    package = 'q2_diversity_lib.tests'
+
+    def setUp(self):
+        super().setUp()
+        self.method = self.plugin.actions['alpha_passthrough']
+        self.available_metrics = alpha.METRICS['NONPHYLO']['UNIMPL']
```
Comment (thermokarst) [inline q2_diversity_lib/tests/test_alpha.py:250]: Seems like unneccessary indirection to me - `METRICS` is already a "package constant", I don't think giving it an alias improves the clarity of the tests in this suite.

File: q2_diversity_lib/tests/test_alpha.py [hunk 9]
```diff
@@ -234,3 +239,42 @@ def test_do_not_drop_undefined_samples(self):
...
+        empty_table = biom.Table(np.array([]), [], [])
+        self.empty_table = Artifact.import_data('FeatureTable[Frequency]',
+                                                empty_table)
+        crawford_tbl = self.get_data_path('crawford.biom')
+        self.crawford_tbl = Artifact.import_data('FeatureTable[Frequency]',
+                                                 crawford_tbl)
+
+    def testMethod(self):
```
Comment (thermokarst) [inline q2_diversity_lib/tests/test_alpha.py:258]: Should this be `test_method`?

```suggestion
    def test_method(self):
```

File: q2_diversity_lib/tests/test_beta.py [hunk 10]
```diff
@@ -321,10 +322,152 @@ def test_accepted_types_have_consistent_behavior(self):
...
+                                                empty_table)
+        crawford_tbl = self.get_data_path('crawford.biom')
+        self.crawford_tbl = Artifact.import_data('FeatureTable[Frequency]',
+                                                 crawford_tbl)
+        table = biom.Table(np.array([[0, 1, 3], [1, 1, 2]]),
+                           ['O1', 'O2'],
+                           ['S1', 'S2', 'S3'])
+        self.table = Artifact.import_data('FeatureTable[Frequency]', table)
+
+    def testMethod(self):
```
Comment (thermokarst) [inline q2_diversity_lib/tests/test_beta.py:352]: Should this be `test_method`?

```suggestion
    def test_method(self):
```

File: q2_diversity_lib/tests/test_beta.py [hunk 11]
```diff
@@ -321,10 +322,152 @@ def test_accepted_types_have_consistent_behavior(self):
...
+        self.empty_table = Artifact.import_data('FeatureTable[Frequency]',
+                                                empty_table)
+        crawford_tbl = self.get_data_path('crawford.biom')
+        self.crawford_tbl = Artifact.import_data('FeatureTable[Frequency]',
+                                                 crawford_tbl)
+        crawford_tree = self.get_data_path('crawford.nwk')
+        self.crawford_tree = Artifact.import_data('Phylogeny[Rooted]',
+                                                  crawford_tree)
+
+    def testMethod(self):
```
Comment (thermokarst) [inline q2_diversity_lib/tests/test_beta.py:446]: Should this be `test_method`?

```suggestion
    def test_method(self):
```

File: q2_diversity_lib/tests/test_beta.py [hunk 12]
```diff
@@ -321,10 +322,152 @@ def test_accepted_types_have_consistent_behavior(self):
...
+    def test_passed_implemented_metric(self):
+        # beta_passthrough does not provide access to measures that have been
+        # implemented locally
+        for metric in beta.METRICS['NONPHYLO']['IMPL']:
+            with self.assertRaisesRegex(TypeError, f"{metric}.*incompatible"):
+                self.method(table=self.crawford_tbl, metric=metric)
+
+    def test_aitchison(self):
+        actual = self.method(table=self.table, metric='aitchison')
+        actual = actual[0].view(skbio.DistanceMatrix)
```
Comment (thermokarst) [inline q2_diversity_lib/tests/test_beta.py:377]:
```suggestion
        actual, = self.method(table=self.table, metric='aitchison')
        actual = actual.view(skbio.DistanceMatrix)
```

File: q2_diversity_lib/tests/test_beta.py [hunk 13]
```diff
@@ -321,10 +322,152 @@ def test_accepted_types_have_consistent_behavior(self):
...
+
+    def test_canberra_adkins(self):
+        t = biom.Table(np.array([[0, 0], [0, 1], [1, 2]]),
+                       ['O1', 'O2', 'O3'],
+                       ['S1', 'S2'])
+        t = Artifact.import_data('FeatureTable[Frequency]', t)
+        d = (1. / 2.) * sum([abs(0. - 1.) / (0. + 1.),
+                             abs(1. - 2.) / (1. + 2.)])
+        expected = skbio.DistanceMatrix(np.array([[0.0, d], [d, 0.0]]),
+                                        ids=['S1', 'S2'])
```
Comment (thermokarst) [inline q2_diversity_lib/tests/test_beta.py:396]: Let's not re-implement the metric logic here - instead, please just fill in the expected values in the `expected` DM (like in the jensenshannon test below).

File: q2_diversity_lib/tests/test_beta.py [hunk 14]
```diff
@@ -321,10 +322,152 @@ def test_accepted_types_have_consistent_behavior(self):
...
+        actual = self.method(table=t, metric='canberra_adkins')
+        actual = actual[0].view(skbio.DistanceMatrix)
```
Comment (thermokarst) [inline q2_diversity_lib/tests/test_beta.py:398]:
```suggestion
        actual, = self.method(table=t, metric='canberra_adkins')
        actual = actual.view(skbio.DistanceMatrix)
```

File: q2_diversity_lib/tests/test_util.py [hunk 15]
```diff
@@ -17,7 +17,12 @@
 from q2_types.feature_table import BIOMV210Format
 from q2_types.tree import NewickFormat
 from .._util import (_disallow_empty_tables,
-                     _validate_requested_cpus)
+                     _validate_requested_cpus,
+                     translate_metric_name)
+import q2_diversity_lib
+
+a_METRICS = q2_diversity_lib.alpha.METRICS
+b_METRICS = q2_diversity_lib.beta.METRICS
```
Comment (thermokarst) [inline q2_diversity_lib/tests/test_util.py:25]: Since constants are all uppercase in pep8-land, let's keep it consistent:

```suggestion
A_METRICS = q2_diversity_lib.alpha.METRICS
B_METRICS = q2_diversity_lib.beta.METRICS
```

File: q2_diversity_lib/beta.py [hunk 16]
```diff
@@ -6,17 +6,114 @@
...
+    if metric == 'aitchison':
+        counts += pseudocount
+        metric = aitchison
+    elif metric == 'canberra_adkins':
+        metric = canberra_adkins
+    elif metric == 'jensenshannon':
+        metric = jensen_shannon
+    return skbio.diversity.beta_diversity(
+            metric=metric, counts=counts, ids=sample_ids, validate=True,
+            pairwise_func=sklearn.metrics.pairwise_distances, n_jobs=n_jobs)
```
[the remark this answers is not in the document]
Comment (ChrisKeefe) [reply q2_diversity_lib/beta.py:83]: Thanks. Having the return glued to the end of that if/else block was super gross!

File: q2_diversity_lib/tests/test_alpha.py (hunk 8 above)
Comment (ChrisKeefe) [reply q2_diversity_lib/tests/test_alpha.py:250]: handled here, in `test_beta`, and in `test_beta_phylogenetic`

File: q2_diversity_lib/tests/test_alpha.py (hunk 7 above)
Comment (ChrisKeefe) [reply q2_diversity_lib/tests/test_alpha.py:36]: Am I right that it's the `metric_name_translations` variable you don't like here, not the `translate_metric_names` function?

I'm good with swapping that out, but think the function helps readability a lot.

**Comment by ChrisKeefe** [comment]
Sorry these review-comment commits are broken up a little funny, @thermokarst. Github hid a bunch of "additional comments" when I did my first pass, and I'm disinclined to re-organize these and force-push. 

Unless I misinterpreted the one unresolved comment, this should address everything.

File: q2_diversity_lib/tests/test_alpha.py (hunk 7 above)
Comment (ChrisKeefe) [reply q2_diversity_lib/tests/test_alpha.py]: Oooh! Looks like it was `translate_metric_names()` after all? https://github.com/qiime2/q2-diversity/pull/281#discussion_r471161680

Comment (thermokarst) [reply q2_diversity_lib/tests/test_alpha.py]: Thanks for checking - its the function I'm concerned with - I think having the metrics "automatically" initialized if not listed in the dict might cause some headaches later. Adding a new metric, for example, what if I misspell it? This util will "allow" the misspelling to leak through. Not the end of the world, but then you're stuck debugging with some indirection in the way.

I'm also thinking about the _Zen of Python_ here: https://www.python.org/dev/peps/pep-0020/

> Explicit is better than implicit.
Simple is better than complex.

Comment (ChrisKeefe) [reply q2_diversity_lib/tests/test_alpha.py]: Resolved in 56b88f3.

Comment (thermokarst) [reply q2_diversity_lib/tests/test_alpha.py]: Sorry, just noticed that you edited this before I had responded - somehow I had a stale version of https://github.com/qiime2/q2-diversity-lib/pull/24#discussion_r471812735.

> How do you feel about this?

I like that plan! I agree, the unimplemented metrics don't need to be in the dict.

Looks like 56b88f3 goes with the full list of unimpl (if I am reading right), so your call on the programmatic unimpl changes. Again, real sorry about the stale comment. That's one downside to retroactive comment editing - no notifications!

Comment (ChrisKeefe) [reply q2_diversity_lib/tests/test_alpha.py]: resolved in https://github.com/qiime2/q2-diversity/pull/281/commits/a544885a85510aa04a0f3efeeebe500ccdf6291d and 5a6a620f070cbf02cecab6ef5ef3fc58043d2e3c

File: q2_diversity_lib/tests/test_beta.py [hunk 17]
```diff
@@ -337,7 +337,6 @@ class BetaPassthroughTests(TestPluginBase):
     def setUp(self):
         super().setUp()
         self.method = self.plugin.actions['beta_passthrough']
-        beta.METRICS['NONPHYLO']['UNIMPL'] = beta.METRICS['NONPHYLO']['UNIMPL']
```
Comment (ChrisKeefe) [inline q2_diversity_lib/tests/test_beta.py]: Thanks for catching this!
